### ISSUE TITLE
Fine-grained admin permission client manage does not work

### ISSUE THREAD

**Issue by david-hanger** [keycloak/keycloak#20431]
### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

admin/fine-grained-permissions

### Describe the bug

Tested in both versions 21.1.1 and 20.0.3, the client **manage** permission does not work as documented, from the feature **admin-fine-grained-authz**.  An authorized user can see the client (given the additional documented role), and the UI appears to allow editing the client's **General Settings** tab.  But clicking the **Save** button yields a **403 Forbidden** response with only "unknown error".  The Keycloak logs do not emit any details about the incident.

### Version

21.1.1

### Expected behavior

Following the [documentation for fine-grained client admin privileges](https://www.keycloak.org/docs/21.1.1/server_admin/#managing-one-specific-client) should result in the behavior described in the documentation, in this case specifically allowing the assigned user the capability to edit settings for the target client.  This apparent bug may be specific to the client **manage** and **configure** permissions, because the map-roles and map-roles-composite [permissions work as documented](https://www.keycloak.org/docs/21.1.1/server_admin/#restrict-user-role-mapping).

### Actual behavior

Following the [documentation for fine-grained client admin privileges](https://www.keycloak.org/docs/21.1.1/server_admin/#managing-one-specific-client) results in a user who can view the client settings and modify those settings in the UI but who cannot save those settings, receiving instead a "403 Forbidden" with the message "unknown error".

### How to Reproduce?

1. Stand up MariaDB
    - Container image **docker.io/bitnami/mariadb-galera:10.10.2**
    - We use a 3-node cluster
    - Can provide additional baseline settings if needed, but they shouldn't be relevant
1. Stand up Keycloak
    - Container image **quay.io/keycloak/keycloak:21.1.1**
    - We use a 3-node cluster behind a TLS-terminating load balancer
    - `start --features=admin-fine-grained-authz`
    - Environment variables, excluding sensitive values:
        ```
        KC_CACHE_STACK=ec2
        KC_DB=mariadb
        KC_HEALTH_ENABLED=true
        KC_HOSTNAME_STRICT=false
        KC_HOSTNAME_STRICT_HTTPS=false
        KC_HTTP_ENABLED=true
        KC_PROXY=edge
        KC_SPI_TRUSTSTORE_FILE_FILE=/opt/keycloak/conf/truststore.jks
        KC_TRANSACTION_XA_ENABLED=false
        PROXY_ADDRESS_FORWARDING=true
        ```
1. Create a realm
1. Create a client
1. Create a user
1. Follow the documentation for [managing one specific client](https://www.keycloak.org/docs/21.1.1/server_admin/#managing-one-specific-client)
    - Proceed all the way through that section, ending with *This admin is now able to manage this one client.*
1. Open the client settings (using the just-assigned restricted admin user)
1. Make changes (e.g. edit the **Description** field)
1. Click **Save** at the bottom
1. Observe the error

### Anything else?

Our team is trying to provide developers on our platform with the capability to administer their own team's client settings, but without the ability to affect any other team's clients.  We upgraded from Keycloak 18 to Keycloak 20 to specifically add this fine-grained admin capability, but it does not do what our users need it to do.  Our workaround is to have our multiple dozen development teams contact us for any change they need to make to any of their clients, which quickly bogs our small team down with busywork.  Having this feature work correctly would be a huge load off our proverbial backs.

**Comment by david-hanger** [comment]
It seems rather unusual to me that this is an issue that has persisted across 2 major versions, where the documented steps do not work as described.  I image that having a single-client-admin feature is fairly common, if you chose it as the first example of fine-grained admin controls.  Perhaps there is a step missing to make it work correctly, but I've spent a few hours playing around without being able to make any progress.  The only thing that makes the client edit **Save** correctly is adding to the user the **realm-management** client role **manage-clients**, which of course allows the user to edit all clients.

**Comment by ssilvert** [comment]
@david-hanger Just to be clear, this also did not work on Keycloak 18 with the old admin console?

**Comment by david-hanger** [comment]
My first task with Keycloak after joining this team was upgrading it to v20, so I don't actually know.  This capability was purportedly the primary drive for upgrading from v18 to v20.  I just fired up a test cluster yesterday, so I'll drop that to 18 and see what it does.

**Comment by ssilvert** [comment]
BTW, I tried out your use case and it seems to be working fine with both the old and new consoles.  I don't know if the docs may be wrong, but it's relatively easy to configure a user that only has access to one client.

1. Give the user a role mapping of realm-management-->query-clients
2. Create a policy that maps to the users you want to allow.  The `User` policy type works well for this.
3. Go to the permissions page for the client.
4. Enable permissions
5. Click on `manage`
6. Assign the policy you created in step 2.
7. When the user logs in, he will only see the client from step 3.

**Comment by david-hanger** [comment]
@ssilvert, unfortunately your steps exactly as describe result in the same error.  In case my original submission was not clear, that user is **successfully** able to:
- See the admin console
- See the **Clients** option under **Manage** (in the left panel)
- See only the target client in the list (in the main/right panel)
- Open the target client
- See its settings (under most tabs)
- Modify field values
- See a **Save** button

That user is **not able** to:
- **Save** the changes
    - Clicking the **Save** button yields a red error message in the upper-right corner that reads "Client could not be updated: unknown_error"
    - Checking the browser's Network tab in the dev console shows an HTTP 403 Forbidden response with payload `{"error":"unknown_error"}`

**Comment by ssilvert** [comment]
OK.  I understand your issue a little better.  This is definitely a bug on the back end.  The access object returned to the UI shows `manage=true`.  But during the REST call, the permissions are not evaluated correctly and you get a 403.  I'm sure this used to work.  Must be a regression.

**Comment by david-hanger** [comment]
Given that the documentation stops at showing that the user can see the client in the list, I suspect that the relevant test case for this feature just stopped there, as a "confirm the user can see the client" and not "confirm the user can modify the client".  A seemingly unrelated change may have resulted in this capability silently breaking, with the test case still passing because it doesn't quite go far enough.

**Comment by mposolda** [comment]
Preliminary adding to 22 as it looks like a regression.

On the other hand, it is related to "ADmin Fine Grained permission", which is a "preview" feature and hence lower priority. So we may need to postpone.

**Comment by david-hanger** [comment]
@mposolda, I appreciate the attention to finish the triage.

Is your team open to looking at bug fixes provided from outside pull requests?  If so, I might try my hand at it as a personal project on weekends.

Thanks again.

**Comment by ssilvert** [comment]
> @mposolda, I appreciate the attention to finish the triage.
> 
> ~Is your team open to looking at bug fixes provided from outside pull requests?~ **EDIT**: I found your [CONTRIBUTING.md](https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md), so, I may try my hand at this ticket as a personal project on weekends, having several years of Java development experience.
> 
> Thanks again.

Yes!  Good PRs that fix bugs are always welcome.

**Comment by pedroigor** [comment]
@david-hanger @ssilvert I'm not able to reproduce the issue. I can manage and save the client settings without any issues.

I noticed the [documentation](https://www.keycloak.org/docs/21.1.1/server_admin/#managing-one-specific-client) is a bit misleading because it mentions:

```
When first initialized the manage permission does not have any policies associated with it. You will need to create one by going to the policy tab. To get there fast, click on the *Authorization link* shown in the above image. Then click on the policies tab.
```

I can't see any `Authorization link* in the image or when at the administration console. Instead, I need to click on the `Client details` (in the breadcrumb), click on the `Authorization` tab, and finally click on the `Policies` tab to finally create the policy I want to use to restrict access to the `manage` scope. The UX here is worse than the old administration console because users were able to create a policy from that page and get back to where he was before.

**Comment by adityakuchekar** [comment]
The same issue exists wherein an authorised user has to manage only one group but he can see all the groups in the realm. I have followed the entire procedure as mentioned but it did not work.

**Comment by stianst** [comment]
@ssilvert removing from patch release project as there was no milestone/backport labels added. If you want to backport this one please add backport/23.0 (and backport 22.0 if applicable)

**Comment by monzonj** [comment]
hello. Do we have a fix for this issue? still there in version 23.0.1

**Comment by owenJiao** [comment]
I want to know whether this bug is fixed, now I encounter the same issue with keycloak version 22

**Comment by mklassen** [comment]
Hoping this may be useful to someone with a better understanding of the permissions evaluation.

In testing version 25.0.2, I have found that a User Policy does not work properly, but a Role Policy does. I was testing authorizing a client service account to mange its own client. Creating a User Policy with the service account user and assigning that policy to the "configure.permission.client.<UUID>",  "manage.permission.client.<UUID>", and  "view.permission.client.<UUID>" did not permit access. However if I created a client role, assigned the role to the service account user, created a Role Policy base on that role, and then used the Role Policy to allow access to "configure.permission.client.<UUID>",  "manage.permission.client.<UUID>", and  "view.permission.client.<UUID>" then the service account could access and manage its own client. 

I was specifically testing by hitting the `admin/realms/${REALM}/clients?clientId=${CLIENT}` endpoint and getting an empty list until I switched from the User Policy to a Role Policy.

org/keycloak/services/resources/admin/ClientsResource.java line 147 auth.clients().canView(c) evaluated to false with a User Policy and true with a Role Policy.
